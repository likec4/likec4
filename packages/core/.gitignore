# Build outputs
dist/
lib/
*.tgz
builder/
compute-view/
model/
styles/
types/
utils/

# Auto-generated by likec4ops prepack
builder/package.json
compute-view/relationships/package.json
compute-view/package.json
model/package.json
styles/package.json
types/package.json
utils/package.json
