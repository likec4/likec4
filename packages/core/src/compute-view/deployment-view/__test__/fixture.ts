import { map, prop } from 'remeda'
import type { DeploymentRulesBuilderOp } from '../../../builder'
import { Builder } from '../../../builder'
import { LikeC4Model } from '../../../model'
import type { DeploymentView } from '../../../types'
import { withReadableEdges } from '../../utils/with-readable-edges'
import { computeDeploymentView } from '../compute'

const {
  builder: b,
  model: {
    model,
    person,
    system,
    component,
    webapp,
    mobile,
    ...$m
  },
  deployment: {
    deployment,
    env,
    node,
    zone,
    instanceOf,
    ...$d
  },
  views: {
    views,
    deploymentView,
    $rules,
    $include,
    $exclude,
  },
} = Builder.forSpecification({
  elements: {
    person: {
      style: {
        shape: 'person',
      },
    },
    system: {},
    component: {},
    webapp: {
      style: {
        shape: 'browser',
      },
    },
    mobile: {
      style: {
        shape: 'mobile',
      },
    },
  },
  deployments: {
    env: {},
    zone: {},
    node: {},
  },
  tags: ['old', 'next', 'temp'],
  relationships: { 'https': {} },
})

export const builder = b
  .with(
    model(
      person('customer', {
        title: 'Happy Customer',
      }),
      system('cloud').with(
        component('frontend').with(
          webapp('dashboard'),
          mobile('mobile'),
        ),
        component('auth'),
        component('backend').with(
          component('api'),
        ),
        component('media', {
          shape: 'storage',
        }),
      ),
      system('aws').with(
        component('rds', {
          shape: 'storage',
        }),
        component('s3', {
          shape: 'storage',
        }),
      ),
      system('email'),
    ),
    model(
      $m.rel('customer', 'cloud', 'uses services'),
      $m.rel('customer', 'cloud.frontend.mobile', 'opens mobile app'),
      $m.rel('customer', 'cloud.frontend.dashboard', {
        title: 'opens in browser',
        tags: ['old'],
      }),
      $m.rel('cloud.frontend.dashboard', 'cloud.auth', {
        title: 'authenticates',
        color: 'green',
      }),
      $m.rel('cloud.frontend.dashboard', 'cloud.backend.api', {
        title: 'fetches data',
        tags: ['old'],
      }),
      $m.rel('cloud.frontend.dashboard', 'cloud.media', {
        title: 'fetches media',
        tags: ['temp'],
      }),
      $m.rel('cloud.frontend.mobile', 'cloud.auth', {
        title: 'authenticates',
        color: 'amber',
      }),
      $m.rel('cloud.frontend.mobile', 'cloud.backend.api', 'fetches data'),
      $m.rel('cloud.frontend.mobile', 'cloud.media', 'fetches media'),
      $m.rel('cloud.backend.api', 'cloud.auth', 'authorizes'),
      $m.rel('cloud.backend.api', 'cloud.media', 'uploads media'),
      $m.rel('cloud.backend.api', 'aws.rds', 'reads/writes'),
      $m.rel('cloud.backend.api', 'email', 'sends emails'),
      $m.rel('cloud.media', 'aws.s3', 'uploads'),
      $m.rel('email', 'customer', 'sends emails'),
    ),
    deployment(
      node('customer').with(
        instanceOf('instance', 'customer'),
      ),
      env('prod').with(
        zone('eu').with(
          zone('zone1').with(
            instanceOf('api', 'cloud.backend.api'),
            instanceOf('ui', 'cloud.frontend.dashboard'),
          ),
          zone('zone2').with(
            instanceOf('api', 'cloud.backend.api'),
            instanceOf('ui', 'cloud.frontend.dashboard'),
          ),
          instanceOf('auth', 'cloud.auth'),
          instanceOf('media', 'cloud.media'),
          instanceOf('db', 'aws.rds'),
        ),
        zone('us').with(
          zone('zone1').with(
            instanceOf('api', 'cloud.backend.api'),
            instanceOf('ui', 'cloud.frontend.dashboard'),
          ),
          instanceOf('db', 'aws.rds'),
        ),
      ),
      env('dev').with(
        node('devCustomer').with(
          instanceOf('instance', 'customer'),
        ),
        node('devCloud').with(
          instanceOf('instance', 'cloud'),
        ),
      ),
      env('acc').with(
        node('testCustomer').with(
          instanceOf('instance', 'customer'),
        ),
        zone('eu').with(
          instanceOf('api', 'cloud.backend.api'),
          instanceOf('ui', 'cloud.frontend.dashboard'),
          instanceOf('auth', 'cloud.auth'),
          instanceOf('db', 'aws.rds'),
        ),
      ),
      env('global').with(
        instanceOf('email', 'email'),
      ),
      $d.rel('prod.eu.db', 'prod.us.db', {
        title: 'replicates',
        color: 'green',
      }),
      $d.rel('prod.us.db', 'prod.eu.db', {
        title: 'replicates',
        color: 'amber',
      }),
    ),
  )
  .clone()

export type Types = typeof builder['Types']

export { $exclude, $include }

export function computeView(...rules: DeploymentRulesBuilderOp<Types>[]) {
  const modelsource = builder.clone()
    .with(
      views(
        deploymentView('index', $rules(...rules)),
      ),
    ).build()

  const model = LikeC4Model.create({
    ...modelsource,
    views: {},
  })

  const view = withReadableEdges(computeDeploymentView(model, modelsource.views.index as DeploymentView))

  return Object.assign(view, {
    nodeIds: map(view.nodes, prop('id')) as string[],
    edgeIds: map(view.edges, prop('id')) as string[],
  })
}

export function computeNodesAndEdges(...rules: DeploymentRulesBuilderOp<Types>[]) {
  const { nodeIds, edgeIds } = computeView(...rules)
  return {
    // Starts with capital letter to be first in snapshot
    Nodes: nodeIds,
    edges: edgeIds,
  }
}

export function createModel() {
  const modelsource = builder.build()

  return LikeC4Model.create({
    ...modelsource,
    views: {},
  })
}
