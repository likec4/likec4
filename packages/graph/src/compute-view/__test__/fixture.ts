import { indexBy, values } from 'remeda'
import type {
  ComputedView,
  Element,
  ElementKind,
  Expression as C4Expression,
  Fqn,
  Relation,
  RelationID,
  Tag,
  ViewID,
  ViewRule,
  ViewRuleExpression
} from '@likec4/core'
import { LikeC4ModelGraph } from '../../LikeC4ModelGraph'
import { computeElementView } from '../index'
import { pluck } from 'rambdax'

/**
              ┌──────────────────────────────────────────────────┐
              │                      cloud                       │
              │  ┌───────────────────────────────────────────┐   │
              │  │                 frontend                  │   │
┏━━━━━━━━━━┓  │  │   ┏━━━━━━━━━━━━━┓   ┏━━━━━━━━━━━━━━━━┓    │   │   ┏━━━━━━━━━━━┓
┃          ┃  │  │   ┃             ┃   ┃                ┃    │   │   ┃           ┃
┃ customer ┃──┼──┼──▶┃  dashboard  ┃   ┃   adminpanel   ┃◀───┼───┼───┃  support  ┃
┃          ┃  │  │   ┃             ┃   ┃                ┃    │   │   ┃           ┃
┗━━━━━━━━━━┛  │  │   ┗━━━━━━┳━━━━━━┛   ┗━━━━━━━━┳━━━━━━━┛    │   │   ┗━━━━━━━━━━━┛
              │  └──────────┼───────────────────┼────────────┘   │
              │             ├───────────────────┘                │
              │             │                                    │
              │  ┌──────────┼────────────────────────────────┐   │
              │  │          ▼       backend                  │   │
              │  │   ┏━━━━━━━━━━━━━┓       ┏━━━━━━━━━━━━━┓   │   │
              │  │   ┃             ┃       ┃             ┃   │   │
              │  │   ┃  graphlql   ┃──────▶┃   storage   ┃   │   │
              │  │   ┃             ┃       ┃             ┃   │   │
              │  │   ┗━━━━━━━━━━━━━┛       ┗━━━━━━┳━━━━━━┛   │   │
              │  └────────────────────────────────┼──────────┘   │
              └───────────────────────────────────┼──────────────┘
                                                  │
                                        ┌─────────┼─────────┐
                                        │ amazon  │         │
                                        │         ▼         │
                                        │ ┏━━━━━━━━━━━━━━┓  │
                                        │ ┃              ┃  │
                                        │ ┃      s3      ┃  │
                                        │ ┃              ┃  │
                                        │ ┗━━━━━━━━━━━━━━┛  │
                                        └───────────────────┘

specification {
  element actor
  element system
  element container
  element component

  tag old
}

model {

  actor customer
  actor support

  system cloud {
    container backend {
      component graphql
      component storage {
        #old
      }

      graphql -> storage
    }

    container frontend {
      component dashboard {
        -> graphql
      }
      component adminPanel {
        #old
        -> graphql
      }
    }
  }

  customer -> dashboard
  support -> adminPanel

  system amazon {
    component s3

    cloud.backend.storage -> s3
  }

}

 */
const el = ({
  id,
  kind,
  title,
  ...props
}: Partial<Omit<Element, 'id' | 'kind'>> & { id: string; kind: string }): Element => ({
  id: id as Fqn,
  kind: kind as ElementKind,
  title: title ?? id,
  description: null,
  technology: null,
  tags: null,
  links: null,
  ...props
})

export const fakeElements = {
  'customer': el({
    id: 'customer',
    kind: 'actor',
    title: 'customer',
    shape: 'person'
  }),
  'support': el({
    id: 'support',
    kind: 'actor',
    title: 'support',
    shape: 'person'
  }),
  'cloud': el({
    id: 'cloud',
    kind: 'system',
    title: 'cloud'
  }),
  'amazon': el({
    id: 'amazon',
    kind: 'system',
    title: 'amazon'
  }),
  'amazon.s3': el({
    id: 'amazon.s3',
    kind: 'component',
    title: 's3',
    shape: 'storage'
  }),
  'cloud.backend': el({
    id: 'cloud.backend',
    kind: 'container',
    title: 'backend'
  }),
  'cloud.frontend': el({
    id: 'cloud.frontend',
    kind: 'container',
    title: 'frontend',
    shape: 'browser'
  }),
  'cloud.backend.graphql': el({
    id: 'cloud.backend.graphql',
    kind: 'component',
    title: 'graphql'
  }),
  'cloud.backend.storage': el({
    id: 'cloud.backend.storage',
    kind: 'component',
    title: 'storage',
    tags: ['old' as Tag]
  }),
  'cloud.frontend.adminPanel': el({
    id: 'cloud.frontend.adminPanel',
    kind: 'component',
    title: 'adminPanel',
    tags: ['old' as Tag]
  }),
  'cloud.frontend.dashboard': el({
    id: 'cloud.frontend.dashboard',
    kind: 'component',
    title: 'dashboard'
  })
} satisfies Record<string, Element>

export type FakeElementIds = keyof typeof fakeElements

const rel = ({
  source,
  target,
  title
}: {
  source: FakeElementIds
  target: FakeElementIds
  title?: string
}): Relation => ({
  id: `${source}:${target}` as RelationID,
  title: title ?? '',
  source: source as Fqn,
  target: target as Fqn
})

export const fakeRelations = [
  rel({
    source: 'customer',
    target: 'cloud.frontend.dashboard',
    title: 'opens in browser'
  }),
  rel({
    source: 'support',
    target: 'cloud.frontend.adminPanel',
    title: 'manages'
  }),
  rel({
    source: 'cloud.backend.storage',
    target: 'amazon.s3',
    title: 'uploads'
  }),
  rel({
    source: 'cloud.backend.graphql',
    target: 'cloud.backend.storage',
    title: 'stores'
  }),
  rel({
    source: 'cloud.frontend.dashboard',
    target: 'cloud.backend.graphql',
    title: 'requests'
  }),
  rel({
    source: 'cloud.frontend.adminPanel',
    target: 'cloud.backend.graphql',
    title: 'fetches'
  })
]

export type FakeRelationIds = keyof typeof fakeRelations

export const fakeModel = new LikeC4ModelGraph({
  elements: fakeElements,
  relations: indexBy(fakeRelations, r => r.id)
})

const emptyView = {
  id: 'index' as ViewID,
  title: null,
  description: null,
  tags: null,
  links: null,
  rules: []
}

export const includeWildcard = {
  include: [
    {
      wildcard: true
    }
  ]
} satisfies ViewRule

export type ElementRefExpr = '*' | FakeElementIds | `${FakeElementIds}.*`

type InOutExpr = `-> ${ElementRefExpr} ->`
type IncomingExpr = `-> ${ElementRefExpr}`
type OutgoingExpr = `${ElementRefExpr} ->`
type RelationExpr = `${ElementRefExpr} -> ${ElementRefExpr}`

type Expression = ElementRefExpr | InOutExpr | IncomingExpr | OutgoingExpr | RelationExpr

function toExpression(expr: Expression): C4Expression {
  if (expr === '*') {
    return { wildcard: true }
  }
  if (expr.startsWith('->')) {
    if (expr.endsWith('->')) {
      return {
        inout: toExpression(expr.replace(/->/g, '').trim() as ElementRefExpr) as any
      }
    }
    return {
      incoming: toExpression(expr.replace('-> ', '') as ElementRefExpr) as any
    }
  }
  if (expr.endsWith(' ->')) {
    return {
      outgoing: toExpression(expr.replace(' ->', '') as ElementRefExpr) as any
    }
  }
  if (expr.includes(' -> ')) {
    const [source, target] = expr.split(' -> ')
    return {
      source: toExpression(source as ElementRefExpr) as any,
      target: toExpression(target as ElementRefExpr) as any
    }
  }
  if (expr.endsWith('.*')) {
    return {
      element: expr.replace('.*', '') as Fqn,
      isDescedants: true
    }
  }
  return {
    element: expr as Fqn,
    isDescedants: false
  }
}

export function $include(expr: Expression): ViewRuleExpression {
  return {
    include: [toExpression(expr)]
  }
}
export function $exclude(expr: Expression): ViewRuleExpression {
  return {
    exclude: [toExpression(expr)]
  }
}

export function computeView(
  ...args: [FakeElementIds, ViewRule | ViewRule[]] | [ViewRule | ViewRule[]]
) {
  let result: ComputedView
  if (args.length === 1) {
    result = computeElementView(
      {
        ...emptyView,
        rules: [args[0]].flat()
      },
      fakeModel
    )
  } else {
    result = computeElementView(
      {
        ...emptyView,
        id: 'index' as ViewID,
        viewOf: args[0] as Fqn,
        rules: [args[1]].flat()
      },
      fakeModel
    )
  }
  return Object.assign(result, {
    nodeIds: pluck('id', result.nodes) as string[],
    edgeIds: pluck('id', result.edges) as string[]
  })
}
