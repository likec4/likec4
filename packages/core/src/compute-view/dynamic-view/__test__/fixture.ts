import { partition } from 'remeda'
import type {
  DynamicViewIncludeRule,
  DynamicViewRule,
  DynamicViewStep,
  ElementViewPredicate,
  Fqn,
  ParsedDynamicView as DynamicView,
  ViewId,
} from '../../../types'
import { type $Aux, type FakeElementIds, fakeModel } from '../../element-view/__test__/fixture'
import { computeDynamicView } from '../compute'

const emptyView = {
  _type: 'dynamic' as const,
  id: 'index' as ViewId,
  title: null,
  description: null,
  tags: null,
  links: null,
  rules: [],
}

type StepExpr = `${FakeElementIds} ${'->' | '<-'} ${FakeElementIds}`
type StepProps = Omit<DynamicViewStep, 'source' | 'target' | 'isBackward'>

export function $step(expr: StepExpr, props?: string | Partial<StepProps>): DynamicViewStep {
  const title = typeof props === 'string' ? props : props?.title
  if (expr.includes(' -> ')) {
    const [source, target] = expr.split(' -> ')
    return {
      source: source as Fqn,
      target: target as Fqn,
      ...(typeof props === 'object' ? props : {}),
      title: title ?? null,
    }
  }
  if (expr.includes(' <- ')) {
    const [target, source] = expr.split(' <- ')
    return {
      source: source as Fqn,
      target: target as Fqn,
      ...(typeof props === 'object' ? props : {}),
      title: title ?? null,
      isBackward: true,
    }
  }
  throw new Error(`Invalid step expression: ${expr}`)
}

export function compute(
  stepsAndRules: (DynamicViewStep<$Aux> | ElementViewPredicate<$Aux> | DynamicViewIncludeRule<$Aux>)[],
) {
  const [steps, rules] = partition(stepsAndRules, (s): s is DynamicViewStep => 'source' in s)
  let view = computeDynamicView(
    fakeModel,
    {
      ...emptyView,
      steps,
      rules: rules as DynamicViewRule[],
    } as DynamicView,
  )
  return Object.assign(view, {
    nodeIds: view.nodes.map((node) => node.id) as string[],
    edgeIds: view.edges.map((edge) => edge.id) as string[],
  })
}
