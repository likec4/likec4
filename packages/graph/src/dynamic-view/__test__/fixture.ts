import type { DynamicViewRule, DynamicViewStep, Fqn, ViewID, ViewRuleExpression } from '@likec4/core'
import { partition } from 'remeda'
import { type FakeElementIds, fakeModel } from '../../compute-view/__test__/fixture'
import { computeDynamicView } from '../index'

const emptyView = {
  __: 'dynamic' as const,
  id: 'index' as ViewID,
  title: null,
  description: null,
  tags: null,
  links: null,
  rules: []
}

type StepExpr = `${FakeElementIds} ${'->' | '<-'} ${FakeElementIds}`

export function $step(expr: StepExpr, title?: string): DynamicViewStep {
  if (expr.includes(' -> ')) {
    const [source, target] = expr.split(' -> ')
    return {
      source: source as Fqn,
      target: target as Fqn,
      title: title ?? null
    }
  }
  if (expr.includes(' <- ')) {
    const [target, source] = expr.split(' <- ')
    return {
      source: source as Fqn,
      target: target as Fqn,
      title: title ?? null,
      isBackward: true
    }
  }
  throw new Error(`Invalid step expression: ${expr}`)
}

export function compute(stepsAndRules: (DynamicViewStep | ViewRuleExpression)[]) {
  const [steps, rules] = partition(stepsAndRules, (s): s is DynamicViewStep => 'source' in s)
  const result = computeDynamicView(
    {
      ...emptyView,
      steps,
      rules: rules as DynamicViewRule[]
    },
    fakeModel
  )
  if (!result.isSuccess) {
    throw result.error
  }
  return Object.assign(result.view, {
    nodeIds: result.view.nodes.map((node) => node.id) as string[],
    edgeIds: result.view.edges.map((edge) => edge.id) as string[]
  })
}
