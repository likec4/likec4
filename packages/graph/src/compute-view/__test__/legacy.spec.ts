import { type ElementKind, type Fqn, type IconUrl, type Tag } from '@likec4/core'
import { view } from 'rambdax'
import { describe, expect, it } from 'vitest'
import { $exclude, $include, computeView } from './fixture'

describe('compute-element-view', () => {
  it('should be empty if no root and no rules', () => {
    const { nodes, edges } = computeView([])
    expect(nodes).toEqual([])
    expect(edges).toEqual([])
  })

  it('should show only root if no rules', () => {
    const { nodeIds, edges } = computeView('cloud', [])
    expect(nodeIds).toEqual(['cloud'])
    expect(edges).toEqual([])
  })

  it('should return landscape view on top `include *`', () => {
    const { nodes, nodeIds, edgeIds } = computeView([{ include: [{ wildcard: true }] }])
    expect(nodeIds).toEqual(['support', 'customer', 'cloud', 'amazon'])
    expect(edgeIds).toEqual(['customer:cloud', 'support:cloud', 'cloud:amazon'])
    const [support, customer, cloud, amazon] = nodes
    expect(amazon).toMatchObject({
      outEdges: [],
      inEdges: ['cloud:amazon']
    })
    expect(cloud).toMatchObject({
      outEdges: ['cloud:amazon'],
      inEdges: expect.arrayContaining(['support:cloud', 'customer:cloud'])
    })
    expect(customer).toMatchObject({
      outEdges: ['customer:cloud'],
      inEdges: []
    })
    expect(support).toMatchObject({
      outEdges: ['support:cloud'],
      inEdges: []
    })
  })

  it('should return nodes in the same order as was in view', () => {
    const { nodeIds, edgeIds } = computeView([$include('customer'), $include('*')])
    expect(nodeIds).toEqual(['support', 'customer', 'cloud', 'amazon'])
    expect(edgeIds).toEqual(['customer:cloud', 'support:cloud', 'cloud:amazon'])
  })

  it('view of cloud', () => {
    const { nodeIds, edgeIds, ...view } = computeView('cloud', [$include('*')])
    expect(nodeIds).toEqual([
      'support',
      'customer',
      'cloud.frontend',
      'cloud.backend',
      'cloud',
      'amazon'
    ])

    expect(edgeIds).to.have.same.members([
      'cloud.frontend:cloud.backend',
      'cloud.backend:amazon',
      'customer:cloud.frontend',
      'support:cloud.frontend'
    ])

    expect(view).toMatchSnapshot()
  })

  it('view of cloud.backend', () => {
    const { nodeIds, edgeIds } = computeView('cloud.backend', [$include('*'), $include('customer')])

    expect(nodeIds).toEqual([
      'customer',
      'cloud.frontend',
      'cloud.backend.graphql',
      'cloud.backend.storage',
      'cloud.backend',
      'amazon'
    ])

    expect(edgeIds).to.have.same.members([
      'cloud.backend.graphql:cloud.backend.storage',
      'cloud.backend.storage:amazon',
      'cloud.frontend:cloud.backend.graphql',
      'customer:cloud.frontend'
    ])
  })

  it('view of cloud.frontend', () => {
    const { nodeIds, edgeIds } = computeView('cloud.frontend', [$include('*')])
    expect(nodeIds).toEqual([
      'customer',
      'cloud.frontend.dashboard',
      'support',
      'cloud.frontend.adminPanel',
      'cloud.frontend',
      'cloud.backend'
    ])

    expect(edgeIds).to.have.same.members([
      'cloud.frontend.adminPanel:cloud.backend',
      'cloud.frontend.dashboard:cloud.backend',
      'customer:cloud.frontend.dashboard',
      'support:cloud.frontend.adminPanel'
    ])

    expect(view).toMatchSnapshot()
  })

  it('view of amazon', () => {
    const { nodeIds, edgeIds } = computeView('amazon', [$include('*')])

    expect(nodeIds).toEqual(['cloud', 'amazon.s3', 'amazon'])
    expect(edgeIds).toEqual(['cloud:amazon.s3'])
  })

  it('view of cloud.frontend (and include parent cloud)', () => {
    const { edgeIds, nodeIds } = computeView('cloud.frontend', [$include('*'), $include('cloud')])

    expect(nodeIds).toEqual([
      'customer',
      'cloud.frontend.dashboard',
      'support',
      'cloud.frontend.adminPanel',
      'cloud.frontend',
      'cloud.backend',
      'cloud'
    ])

    expect(edgeIds).to.have.same.members([
      'cloud.frontend.adminPanel:cloud.backend',
      'cloud.frontend.dashboard:cloud.backend',
      'customer:cloud.frontend.dashboard',
      'support:cloud.frontend.adminPanel'
    ])
  })

  it('view of cloud.frontend (and include parent cloud)', () => {
    const { edgeIds, nodeIds } = computeView('cloud.frontend', [$include('*'), $include('cloud')])

    expect(nodeIds).toEqual([
      'customer',
      'cloud.frontend.dashboard',
      'support',
      'cloud.frontend.adminPanel',
      'cloud.frontend',
      'cloud.backend',
      'cloud'
    ])

    expect(edgeIds).to.have.same.members([
      'cloud.frontend.adminPanel:cloud.backend',
      'cloud.frontend.dashboard:cloud.backend',
      'customer:cloud.frontend.dashboard',
      'support:cloud.frontend.adminPanel'
    ])
  })

  it('view of cloud.frontend (and include parent cloud)', () => {
    const { edgeIds, nodeIds } = computeView('cloud.frontend', [
      $include('*'),
      $include('cloud'),
      $include('cloud.frontend -> cloud.backend.*')
    ])

    expect(nodeIds).toEqual([
      'customer',
      'cloud.frontend.dashboard',
      'support',
      'cloud.frontend.adminPanel',
      'cloud.frontend',
      'cloud.backend.graphql',
      'cloud.backend',
      'cloud'
    ])

    expect(edgeIds).to.have.same.members([
      'customer:cloud.frontend.dashboard',
      'support:cloud.frontend.adminPanel',
      'cloud.frontend.adminPanel:cloud.backend.graphql',
      'cloud.frontend.dashboard:cloud.backend.graphql'
    ])
  })

  it('view of cloud (exclude cloud, amazon)', () => {
    const { edgeIds, nodeIds } = computeView('cloud', [
      $include('*'),
      $exclude('cloud'),
      $exclude('amazon')
    ])

    expect(nodeIds).toEqual(['support', 'customer', 'cloud.frontend', 'cloud.backend'])

    expect(edgeIds).to.have.same.members([
      'cloud.frontend:cloud.backend',
      'customer:cloud.frontend',
      'support:cloud.frontend'
    ])
  })

  it('view with 3 levels', () => {
    const { edgeIds, nodeIds, ...view } = computeView('cloud', [
      $include('*'),
      $include('cloud.frontend.*'),
      $include('cloud.backend.*'),
      $include('-> amazon.*'),
      $exclude('amazon'),
      $exclude('cloud.frontend')
    ])

    expect(nodeIds).toEqual([
      'customer',
      'cloud.frontend.dashboard',
      'support',
      'cloud.frontend.adminPanel',
      'cloud.backend.graphql',
      'cloud.backend.storage',
      'cloud.backend',
      'cloud',
      'amazon.s3'
    ])

    expect(edgeIds).to.have.same.members([
      'support:cloud.frontend.adminPanel',
      'customer:cloud.frontend.dashboard',
      'cloud.backend.graphql:cloud.backend.storage',
      'cloud.frontend.adminPanel:cloud.backend.graphql',
      'cloud.frontend.dashboard:cloud.backend.graphql',
      'cloud.backend.storage:amazon.s3'
    ])
  })

  it('view with 3 levels (with only relevant elements)', () => {
    const { edgeIds, nodeIds } = computeView('cloud', [
      $include('*'),
      $include('-> cloud.frontend.*'),
      $include('-> cloud.backend.*')
    ])

    expect(nodeIds).toEqual([
      'support',
      'cloud.frontend.adminPanel',
      'customer',
      'cloud.frontend.dashboard',
      'cloud.frontend',
      'cloud.backend.graphql',
      'cloud.backend',
      'cloud',
      'amazon'
    ])

    expect(edgeIds).to.have.same.members([
      'cloud.backend:amazon',
      'customer:cloud.frontend.dashboard',
      'support:cloud.frontend.adminPanel',
      'cloud.frontend.dashboard:cloud.backend.graphql',
      'cloud.frontend.adminPanel:cloud.backend.graphql'
    ])
  })

  it('view of amazon', () => {
    const { edgeIds, nodeIds, ...view } = computeView('amazon', [
      $include('*'),
      $include('cloud.* ->')
    ])

    expect(nodeIds).toEqual(['cloud.backend', 'cloud', 'amazon.s3', 'amazon'])

    expect(edgeIds).to.have.same.members(['cloud.backend:amazon.s3'])
  })

  it('index view with applied styles', () => {
    const { nodes } = computeView([
      $include('customer'),
      $include('amazon'),
      $include('cloud'),
      $include('cloud.frontend'),
      // all elements
      // color: secondary
      {
        targets: [{ wildcard: true }],
        style: {
          color: 'secondary',
          shape: 'storage'
        }
      },
      // cloud
      // color: muted
      {
        targets: [{ element: 'cloud' as Fqn, isDescedants: false }],
        style: {
          color: 'muted',
          icon: 'http://some-icon' as IconUrl
        }
      },
      // cloud.*
      // shape: browser
      {
        targets: [{ element: 'cloud' as Fqn, isDescedants: true }],
        style: {
          shape: 'browser'
        }
      }
    ])

    const amazon = nodes.find(n => n.id === 'amazon')!
    const customer = nodes.find(n => n.id === 'customer')!
    const cloud = nodes.find(n => n.id === 'cloud')!
    const frontend = nodes.find(n => n.id === 'cloud.frontend')!

    expect(amazon).toMatchObject({
      color: 'secondary',
      shape: 'storage'
    })
    expect(amazon).not.toHaveProperty('icon')
    expect(customer).toMatchObject({
      color: 'secondary',
      shape: 'storage'
    })
    expect(customer).not.toHaveProperty('icon')

    expect(cloud).toMatchObject({
      color: 'muted',
      shape: 'storage',
      icon: 'http://some-icon'
    })
    expect(frontend).toMatchObject({
      parent: 'cloud',
      color: 'secondary',
      shape: 'browser'
    })
    expect(frontend).not.toHaveProperty('icon')
  })

  it('should include by element kind', () => {
    const { nodeIds, edgeIds } = computeView({
      include: [
        {
          elementKind: 'system' as ElementKind,
          isEqual: true
        }
      ]
    })

    expect(nodeIds).toEqual(['cloud', 'amazon'])
    expect(edgeIds).toEqual(['cloud:amazon'])
  })

  it('should include by element tag', () => {
    const { nodeIds, edgeIds } = computeView({
      include: [
        {
          elementTag: 'old' as Tag,
          isEqual: true
        }
      ]
    })

    expect(nodeIds).toEqual(['cloud.backend.storage', 'cloud.frontend.adminPanel'])

    expect(edgeIds).toEqual([])
  })

  it('should exclude by element tag and kind', () => {
    const { nodeIds, edgeIds } = computeView('cloud', [
      $include('*'),
      $include('cloud.backend.*'),
      $include('cloud.frontend.*'),
      {
        exclude: [
          {
            elementKind: 'actor' as ElementKind,
            isEqual: true
          },
          {
            elementTag: 'old' as Tag,
            isEqual: true
          }
        ]
      }
    ])

    expect(nodeIds).toEqual([
      'cloud.frontend.dashboard',
      'cloud.frontend',
      'cloud.backend.graphql',
      'cloud.backend',
      'cloud',
      'amazon'
    ])

    expect(edgeIds).to.have.same.members([
      'cloud.backend:amazon',
      'cloud.frontend.dashboard:cloud.backend.graphql'
    ])
  })
})
