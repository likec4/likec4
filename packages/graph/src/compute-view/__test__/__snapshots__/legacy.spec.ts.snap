// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compute-element-view > view of cloud 1`] = `
{
  "autoLayout": "TB",
  "description": null,
  "edges": [
    {
      "id": "cloud.frontend:cloud.backend",
      "label": null,
      "parent": "cloud",
      "relations": [
        "cloud.frontend.dashboard:cloud.backend.graphql",
        "cloud.frontend.adminPanel:cloud.backend.graphql",
      ],
      "source": "cloud.frontend",
      "target": "cloud.backend",
    },
    {
      "id": "cloud.backend:amazon",
      "label": "uploads",
      "parent": null,
      "relations": [
        "cloud.backend.storage:amazon.s3",
      ],
      "source": "cloud.backend",
      "target": "amazon",
    },
    {
      "id": "customer:cloud.frontend",
      "label": "opens in browser",
      "parent": null,
      "relations": [
        "customer:cloud.frontend.dashboard",
      ],
      "source": "customer",
      "target": "cloud.frontend",
    },
    {
      "id": "support:cloud.frontend",
      "label": "manages",
      "parent": null,
      "relations": [
        "support:cloud.frontend.adminPanel",
      ],
      "source": "support",
      "target": "cloud.frontend",
    },
  ],
  "id": "index",
  "links": null,
  "nodes": [
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "support",
      "inEdges": [],
      "kind": "actor",
      "level": 0,
      "links": null,
      "outEdges": [
        "support:cloud.frontend",
      ],
      "parent": null,
      "shape": "person",
      "tags": null,
      "technology": null,
      "title": "support",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "customer",
      "inEdges": [],
      "kind": "actor",
      "level": 0,
      "links": null,
      "outEdges": [
        "customer:cloud.frontend",
      ],
      "parent": null,
      "shape": "person",
      "tags": null,
      "technology": null,
      "title": "customer",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "cloud.frontend",
      "inEdges": [
        "customer:cloud.frontend",
        "support:cloud.frontend",
      ],
      "kind": "container",
      "level": 1,
      "links": null,
      "outEdges": [
        "cloud.frontend:cloud.backend",
      ],
      "parent": "cloud",
      "shape": "browser",
      "tags": null,
      "technology": null,
      "title": "frontend",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "cloud.backend",
      "inEdges": [
        "cloud.frontend:cloud.backend",
      ],
      "kind": "container",
      "level": 1,
      "links": null,
      "outEdges": [
        "cloud.backend:amazon",
      ],
      "parent": "cloud",
      "shape": "rectangle",
      "tags": null,
      "technology": null,
      "title": "backend",
    },
    {
      "children": [
        "cloud.frontend",
        "cloud.backend",
      ],
      "color": "primary",
      "description": null,
      "id": "cloud",
      "inEdges": [],
      "kind": "system",
      "level": 0,
      "links": null,
      "outEdges": [],
      "parent": null,
      "shape": "rectangle",
      "tags": null,
      "technology": null,
      "title": "cloud",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "amazon",
      "inEdges": [
        "cloud.backend:amazon",
      ],
      "kind": "system",
      "level": 0,
      "links": null,
      "outEdges": [],
      "parent": null,
      "shape": "rectangle",
      "tags": null,
      "technology": null,
      "title": "amazon",
    },
  ],
  "tags": null,
  "title": null,
  "viewOf": "cloud",
}
`;

exports[`compute-element-view > view of cloud.frontend 1`] = `[Function]`;
